ModelInterface/ModelGUI2/queries/CostCurveQueryBuilder.java

```diff
@@ -324,4 +327,105 @@ public Map addToDataTree(ANode currNode, Map dataTree, DataPair<String, String>
+        QueryRow currRow = null;
+        if(parentPath.hasNext()) {
+            currRow = parentPath.next();
+            if(currRow.id != null && currRow.id.is(currNode)) {
+                // The rest of this path is the same as the last node proccessed.
+                return;
+            } else {
+                currRow = new QueryRow(currNode);
+                parentPath.set(currRow);
+            }
+        } else {
+            currRow = new QueryRow(currNode);
+            parentPath.add(currRow);
+        }
+        BXNode currDOM = BXNode.get(currNode);
+
+		// cache node properties since these may need to go back to the database which could
+		// be expensive
+		String nodeName = currDOM.getNodeName();
+		// run functions can not utilize an attribute map cache since they rely on getNodeHandle
+		// which is not available in those kinds of queries
+		Map<String, String> attrMap = XMLDB.getAttrMapWithCache(currDOM);
+
+		// set the type as the node name if it does not have a type attribute
+		String type = attrMap.get("type");
+		if(type == null) {
+			type = nodeName;
+		}
+
+		// try to find the axis values at the current node
+		boolean setNodeLevel = false;
+		boolean setYearLevel = false;
+		if(nodeName.equals(qg.nodeLevel.getKey())) {
+			setNodeLevel = true;
+            currRow.key = qg.axis1Name;
+			if(qg.nodeLevel.getKey().equals("UndiscountedCost") || qg.nodeLevel.getKey().equals("DiscountedCost")) {
+				if(isGlobal) {
+					currRow.value = "Global";
+				} else {
+					currRow.value = attrMap.get("name");
+				}
+			} else if(isGlobal) {
+                currRow.value = "Global";
+			} else {
+                currRow.value = currDOM.getFirstChild().getFirstChild().getNodeValue();
+			}
+
+		}
+		if(nodeName.equals(qg.yearLevel.getKey())) {
+			setYearLevel = true;
+            if(setNodeLevel) {
+                currRow = new QueryRow(currNode);
+                if(parentPath.hasNext()) {
+                    parentPath.next();
+                    parentPath.set(currRow);
+                } else {
+                    parentPath.add(currRow);
+                }
+            }
+            currRow.key = qg.axis2Name;
+			if(qg.nodeLevel.getKey().equals("UndiscountedCost") || qg.nodeLevel.getKey().equals("DiscountedCost")) {
+                currRow.value = qg.nodeLevel.getKey();
+		} else {
+            if(qg.axis2Name.equals("Year")) {
+                currRow.value = currDOM.getFirstChild().getFirstChild().getNodeValue();
+            } else {
+                currRow.value = currDOM.getFirstChild().getNextSibling().getFirstChild().getNodeValue();
+            }
+		}
+
+		}
+		// if we should not collapse this node then pick out the attributes which
+		// check if we need to collapse this level
+		if(!setNodeLevel && !setYearLevel && !attrMap.isEmpty()) {
+            // TODO: add the ability to use showAttrMap
+			String attr = XMLDB.getAllAttr(attrMap, showAttrList);
+            if(attr != "") {
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [ES: Comparison of String objects using == or != (ES_COMPARING_STRINGS_WITH_EQ)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#es-comparison-of-string-objects-using-or-es-comparing-strings-with-eq).
This code compares java.lang.String objects for reference equality using the == or != operators. Unless both strings are either constants in a source file, or have been interned using the String.intern() method, the same string value may be represented by two different String objects. Consider using the equals(Object) method instead.